core/shared/src/main/scala/fs2/text.scala

```diff
@@ -188,84 +189,52 @@ object text {
+            stringBuilder.clear()
+          case '\r' if i + 1 < string.size && string(i + 1) == '\n' =>
+            linesBuffer += stringBuilder.result()
+            stringBuilder.clear()
+            i += 1
+          case other =>
+            stringBuilder.append(other)
         }
         i += 1
       }
-      val carry = string.substring(start, string.size)
-      (out, carry)
-    }
-
-    def extractLines(
-        buffer: Vector[String],
-        chunk: Chunk[String],
-        pendingLineFeed: Boolean
-    ): (Chunk[String], Vector[String], Boolean) = {
-      @tailrec
-      def go(
-          remainingInput: Vector[String],
-          buffer: Vector[String],
-          output: Vector[String],
-          pendingLineFeed: Boolean
-      ): (Chunk[String], Vector[String], Boolean) =
-        if (remainingInput.isEmpty)
-          (Chunk.indexedSeq(output), buffer, pendingLineFeed)
-        else {
-          val next = remainingInput.head
-          if (pendingLineFeed)
-            if (next.headOption == Some('\n')) {
-              val out = (buffer.init :+ buffer.last.init).mkString
-              go(next.tail +: remainingInput.tail, Vector.empty, output :+ out, false)
-            } else
-              go(remainingInput, buffer, output, false)
-          else {
-            val (out, carry) = linesFromString(next)
-            val pendingLF =
-              if (carry.nonEmpty) carry.last == '\r' else pendingLineFeed
-            go(
-              remainingInput.tail,
-              if (out.isEmpty) buffer :+ carry else Vector(carry),
-              if (out.isEmpty) output
-              else output ++ ((buffer :+ out.head).mkString +: out.tail),
-              pendingLF
-            )
-          }
-        }
-      go(chunk.toVector, buffer, Vector.empty, pendingLineFeed)
     }
 
     def go(
-        buffer: Vector[String],
-        pendingLineFeed: Boolean,
-        s: Stream[F, String]
+        stream: Stream[F, String],
+        stringBuilder: StringBuilder,
+        first: Boolean
     ): Pull[F, String, Unit] =
-      s.pull.uncons.flatMap {
-        case Some((chunk, s)) =>
-          val (toOutput, newBuffer, newPendingLineFeed) =
-            extractLines(buffer, chunk, pendingLineFeed)
-          Pull.output(toOutput) >> go(newBuffer, newPendingLineFeed, s)
-        case None if buffer.nonEmpty =>
-          Pull.output1(buffer.mkString)
-        case None => Pull.done
+      stream.pull.uncons.flatMap {
+        case None => if (first) Pull.done else Pull.output1(stringBuilder.result())
+        case Some((chunk, stream)) =>
+          val linesBuffer = new mutable.ArrayBuffer[String]()
+          chunk.foreach { string =>
+            fillBuffers(stringBuilder, linesBuffer, string)
+          }
+          Pull.output(Chunk.buffer(linesBuffer)) >> go(stream, stringBuilder, first = false)
       }
 
-    s => go(Vector.empty, false, s).stream
+    s => go(s, new mutable.StringBuilder(), first = true).stream
```

Comment:
I think this needs a `Stream.suspend` in order to prevent a bug where the stream returned from the pull is used multiple times. E.g., `val r = src.through(text.lines); r.toList; r.toList` will likely result in the second call returning twice the number of lines. To fix, `s => Stream.suspend(go(...).stream)`

Reply:
yes, definetely, I'll add it!

I have two concerns about the last commit.
1. Using stringBuilder
2. `def fillBuffers` mutates StringBuilder and resulting ArrayBuffer. Thus, looks extremely antiscala-like.

But it gives us extra +50% of throughput compared to the original solution.

It is also possible to make `def linesFromString(strings: Vector[String]): (ArrayBuffer[String], Vector[String])`. This should be closely perfomant to the last commit, but avoids both concerns. I'll jmh it and then rerequest the review, ok?

Reply:
Jmh of vector-based solution:
```
[info] LinesBenchmark.linesBenchmark                0            4  thrpt    3  78059,578 ± 26980,633  ops/s
[info] LinesBenchmark.linesBenchmark                0           16  thrpt    3  76846,929 ± 18504,105  ops/s
[info] LinesBenchmark.linesBenchmark                0           64  thrpt    3  76316,158 ± 12088,617  ops/s
[info] LinesBenchmark.linesBenchmark                1            4  thrpt    3  16542,145 ±  4163,434  ops/s
[info] LinesBenchmark.linesBenchmark                1           16  thrpt    3  26656,283 ±  4506,271  ops/s
[info] LinesBenchmark.linesBenchmark                1           64  thrpt    3  32159,405 ±  5460,660  ops/s
[info] LinesBenchmark.linesBenchmark               10            4  thrpt    3   4726,447 ±  3595,960  ops/s
[info] LinesBenchmark.linesBenchmark               10           16  thrpt    3  14335,162 ±  4119,956  ops/s
[info] LinesBenchmark.linesBenchmark               10           64  thrpt    3  23150,317 ±  4973,143  ops/s
[info] LinesBenchmark.linesBenchmark              100            4  thrpt    3   688,504 ± 174,908  ops/s
[info] LinesBenchmark.linesBenchmark              100           16  thrpt    3  3106,952 ± 651,462  ops/s
[info] LinesBenchmark.linesBenchmark              100           64  thrpt    3   5734,140 ±  2903,247  ops/s
```